### ISSUE TITLE
self.eps

### ISSUE THREAD

**Issue by Oussamab21** [lmartak/distill-nn-tree#2]
Hi , thanks for sharing this  useful code, I was wondering what is the goal of self.eps attribute for the Soft Tree ?

Thank you

**Comment by lmartak** [comment]
Hi, glad my code could help!

The small epsilon constant `self.eps = 1e-8` is there just for numerical stability reasons, in case the `self.loss_leaves` term that goes into logarithm gets so low that it's truncated to zero by limitations in floating point precision. It's there so that `log(0)` is never encountered in the `self.loss` term, as that would evaluate to `-Inf` which you don't really want to go on calculating your gradients from.

https://github.com/lmartak/distill-nn-tree/blob/7ea4632eb497808b1a70623c48e4aceffb15fcac/models/tree.py#L199-L200

Hope this helps. Good luck!

**Comment by Oussamab21** [comment]
Thank you dude I just had another question where are the weights for the soft decision tree are they embedded in self.dense and they fall in Scale as input[1] ? so what is in input[0] i tried to eval them or sess.run() them by tf but the place holder rises an error 
Thank you

**Comment by lmartak** [comment]
> where are the weights for the soft decision tree are they embedded in self.dense and they fall in Scale as input[1] ?

* no, weights are not in `self.dense`
* `self.dense` is a scalar (or vector if `batch_size > 1`) representing **correlation** of input with given node's kernel matrix (kernel = the actual weights created by `self.dense` layer that you're asking about)
* yes, this correlation falls into `Scale` as `input[0]`
* you can access weights (as demonstrated below) in `tf.keras`y manner, reducing any need to use `session` to `run` or `eval` ops, whatsoever

https://github.com/lmartak/distill-nn-tree/blob/7ea4632eb497808b1a70623c48e4aceffb15fcac/models/utils.py#L35-L38

> so what is in input[0]

* into `input[0]` goes so called **inverse temperature** (denoted in [paper](https://arxiv.org/pdf/1711.09784.pdf) as `\beta`) which is stored in `self.inv_temp`, and is used globally for all leaves across the tree 
* it is either set as a **hyperparameter** or being **learned** along with the kernels, [as suggested by author of the approach](https://twitter.com/nickfrosst/status/1032286254057041920)

In case the code doesn't reproduce results as you can see them in the [jupyter notebook](https://github.com/lmartak/distill-nn-tree/blob/master/mnist.ipynb), you can try regressing back to TF version 1.10.0 on which this code was developed.

But generally speaking, looking at the code in this repo and studying documentation on [tf.keras](https://www.tensorflow.org/guide/keras) and [keras.io](https://keras.io/) shall do enough to answer your further questions of this kind :)

Good luck!
